clcache.py

```diff
@@ -1600,57 +1574,89 @@ def processCompileRequest(cache, compiler, args):
+            return ensureArtifactsExist(cache, artifactSection, cachekey, unusableManifestMissReason,
+                                        objectFile, compilerResult)
+
+        entry = createManifestEntry(manifestHash, includePaths)
+        cachekey = entry.objectHash
 
-                    return getOrSetArtifacts(
-                        cache, cachekey, objectFile, compiler, cmdLine, Statistics.registerEvictedMiss)
-            except IncludeNotFoundException:
-                pass
+        def addManifest():
+            manifest = createOrUpdateManifest(manifestSection, manifestHash, entry)
+            manifestSection.setManifest(manifestHash, manifest)
 
-        return postprocessUnusableManifestMiss(
-            cache, objectFile, manifestSection, manifestHash, sourceFile, compiler, cmdLine,
-            Statistics.registerHeaderChangedMiss)
+        artifactSection = cache.compilerArtifactsRepository.section(cachekey)
+        return ensureArtifactsExist(cache, artifactSection, cachekey, unusableManifestMissReason,
+                                    objectFile, compilerResult, addManifest)
 
 
 def processNoDirect(cache, objectFile, compiler, cmdLine, environment):
     cachekey = CompilerArtifactsRepository.computeKeyNodirect(compiler, cmdLine, environment)
-    return getOrSetArtifacts(cache, cachekey, objectFile, compiler, cmdLine, Statistics.registerCacheMiss, environment)
-
-
-def getOrSetArtifacts(cache, cachekey, objectFile, compiler, cmdLine, statsField, environment=None):
     artifactSection = cache.compilerArtifactsRepository.section(cachekey)
-    cleanupRequired = False
     with artifactSection.lock:
         if artifactSection.hasEntry(cachekey):
             return processCacheHit(cache, objectFile, cachekey)
 
-        compilerResult = invokeRealCompiler(compiler, cmdLine, captureOutput=True, environment=environment)
-        returnCode, compilerStdout, compilerStderr = compilerResult
-        with cache.statistics.lock, cache.statistics as stats:
-            statsField(stats)
-            if returnCode == 0 and os.path.exists(objectFile):
-                artifacts = CompilerArtifacts(objectFile, compilerStdout, compilerStderr)
-                cleanupRequired = addObjectToCache(stats, cache, artifactSection, cachekey, artifacts)
+    compilerResult = invokeRealCompiler(compiler, cmdLine, captureOutput=True, environment=environment)
+
+    return ensureArtifactsExist(cache, artifactSection, cachekey, Statistics.registerCacheMiss,
+                                objectFile, compilerResult)
+
 
-    return compilerResult + (cleanupRequired,)
+def ensureArtifactsExist(cache, section, cachekey, reason, objectFile, compilerResult, extraCallable=None):
+    cleanupRequired = False
+    returnCode, compilerOutput, compilerStderr = compilerResult
+    with section.lock:
+        if not section.hasEntry(cachekey):
+            with cache.statistics.lock, cache.statistics as stats:
+                reason(stats)
```

Comment:
Indeed, there is some inconsistency - good catch! I agree that for now, it's plausible to keep the PR as it is -- but this might be something one might want to address in the future...e